### ISSUE TITLE
Python 3 support

### ISSUE THREAD

**Issue by LenzGr** [SUSE/DeepSea#781]
In the near future, Linux distributions are likely to either deprecate or drop support for Python 2. We should ensure that DeepSea has been properly ported and tested on Python 3. New development should be done with Python 3 support in mind.

**Comment by jschmid1** [comment]
We are currently running `salt-master 2016.11.4 (Carbon)`

according to https://github.com/saltstack/salt/issues/11995#issuecomment-289060670 salt supports python3 starting with the 2017.$x release.

We just enabled pylint which can be configured to raise incidents with py2/3 compatibility.

**Comment by jschmid1** [comment]
From saltstack's 2017 releasenotes:

```
REMEMBER TO UPDATE THE SALT MASTER FIRST
Salt's policy has always been that when upgrading, the minion should never be on a newer version than the master. Specifically with this update, because of changes in the fileclient, the 2017.7 minion requires a 2017.7 master.

Backwards compatiblity is still maintained, so older minions can still be used.
```

**Comment by jschmid1** [comment]
python3 changes the way how compiled bytecode is stored.

py2 saved a .pyc alongside the actual file.

py3 creates a __pycache__ directory.

**Comment by jschmid1** [comment]
python-boto's successor is now called [boto3](https://github.com/boto/boto3) which [claims](https://github.com/boto/boto3/blob/develop/tox.ini#L2) to support python3..

**Comment by jschmid1** [comment]
`platform.linux_distribution` returns an empty tuple in SLE15.. 

from the [docs](https://docs.python.org/3/library/platform.html)

`Deprecated since version 3.5, will be removed in version 3.7.`

**Comment by jschmid1** [comment]
the format of /proc/cmdline was changed from 

``` BOOT_IMAGE=/boot/vmlinuz-4.12.14-11.2-default root=/dev/vda1 disk=/dev/vda nomodeset LOCAL_BOOT=yes ```


to 
``` BOOT_IMAGE=(hd0,msdos1)/boot/vmlinuz-4.12.14-11.2-default root=/dev/vda1 disk=/dev/vda nomodeset LOCAL_BOOT=yes  ```

**Comment by jschmid1** [comment]
`tgt_type` is the new `expr_form` since salt 2017.x

**Comment by jschmid1** [comment]
`imp` is now called `importlib`

**Comment by jschmid1** [comment]
subprocesses now often return a  'bytes' object, which is a binary sequence of bytes rather than a string of Unicode characters (you got that 99% of the time in python2). If the stdout of a subprocess is handled without checking for types, we run into exceptions like:

> *** TypeError: startswith first arg must be bytes or a tuple of bytes, not str

When comparing bytes with strings

We can use 

``` python
(b'hi\n'.decode('ascii'))
```

to avoid this.

It will require [quite some changes ](https://github.com/SUSE/DeepSea/search?utf8=%E2%9C%93&q=Popen&type=)

**Comment by jschmid1** [comment]
the cli fails with a dependency issue:

```
admin:~ # deepsea stage run ceph.stage.1
Traceback (most recent call last):
  File "/usr/bin/deepsea", line 6, in <module>
    from pkg_resources import load_entry_point
ImportError: No module named pkg_resources

```

**Comment by jschmid1** [comment]
log.warn is deprecated (in favor of log.warning)

**Comment by jschmid1** [comment]
sorted's cmp parameter is gone.

from the python [docs](https://docs.python.org/3/howto/sorting.html#the-old-way-using-the-cmp-parameter):

```
Many constructs given in this HOWTO assume Python 2.4 or later.
Before that, there was no sorted() builtin and list.sort() took no keyword arguments.
Instead, all of the Py2.x versions supported a cmp parameter to 
handle user specified comparison functions.

In Py3.0, the cmp parameter was removed entirely 
(as part of a larger effort to simplify and unify the language, eliminating the conflict between rich comparisons and the __cmp__() magic method).
```

That can be worked around with:

``` python
To convert to a key function, just wrap the old comparison function:

>>>
>>> sorted([5, 2, 4, 1, 3], key=cmp_to_key(reverse_numeric))
[5, 4, 3, 2, 1]


In Python 3.2, the functools.cmp_to_key() function was added to the functools module in the standard library.
```

**Comment by smithfarm** [comment]
@jschmid1 Regarding `platform.linux_distribution` the following bug is relevant: http://bugzilla.suse.com/show_bug.cgi?id=997614

tl;dr (AAUI) is: `platform.linux_distribution` should not be broken on SUSE in Python 3.6, but it (currently) is - that's why the bug is open - and the preferred replacement, python-distro, may not be packaged/available yet. 

Conclusion: this one is potentially a blocker for us, though I guess it could be worked around by testing for `('', '', '')` and assuming this means "SUSE" haha.

**Comment by jschmid1** [comment]
> Conclusion: this one is potentially a blocker for us, though I guess it could be worked around by testing for ('', '', '') and assuming this means "SUSE" haha.

it can be worked around by using the 

``` python
__grains__['os']
```

which allows us to get rid of one dependency 🎉

**Comment by jschmid1** [comment]
> pkg_resources.DistributionNotFound: The 'deepsea==0.9.1+61.g8f4fbe145' distribution was not found and is required by the application

raised by the CLI from probably the 'python3-setuptools'

**Comment by jschmid1** [comment]
boto3 related errors in the logs:

```
Feb 07 10:35:35 mon1 salt-minion[1733]: [ERROR   ] Exception raised when processing __virtual__ function for salt.loaded.int.module.boto_elbv2. Module will not be loaded: 'boto3.assign_funcs'
Feb 07 10:35:35 mon1 salt-minion[1733]: [WARNING ] salt.loaded.int.module.boto_elbv2.__virtual__() is wrongly returning `None`. It should either return `True`, `False` or a new name. If you're the developer of the module 'boto_elbv2', please fix this.
Feb 07 10:35:35 mon1 salt-minion[1733]: [ERROR   ] Exception raised when processing __virtual__ function for salt.loaded.int.module.boto_kinesis. Module will not be loaded: 'boto3.assign_funcs'
Feb 07 10:35:35 mon1 salt-minion[1733]: [WARNING ] salt.loaded.int.module.boto_kinesis.__virtual__() is wrongly returning `None`. It should either return `True`, `False` or a new name. If you're the developer of the module 'boto_kinesis', please fix this.

```

**Comment by jschmid1** [comment]
the installation for python3-boto3 fails with

```
admin:/var/log/zypp # zypper in python3-boto3
Loading repository data...
Reading installed packages...
Resolving package dependencies...

Problem: nothing provides python3-botocore >= 1.8.0 needed by python3-boto3-1.5.1-1.4.noarch
 Solution 1: do not install python3-boto3-1.5.1-1.4.noarch
 Solution 2: break python3-boto3-1.5.1-1.4.noarch by ignoring some of its dependencies

Choose from above solutions by number or cancel [1/2/c] (c): 

```

**Comment by jschmid1** [comment]
Salt Sodium will deprecate module.run (to what?)
```
 [WARNING ] The function "module.run" is using its deprecated version and will expire in version "Sodium".
```

**Comment by smithfarm** [comment]
@jschmid1 Regarding python3-boto3 installation failure, in which operating system are you seeing this? (Please forgive me if you already wrote that - I'm too lazy to re-read all the comments)

**Comment by jschmid1** [comment]
@smithfarm that's SLE15 and is already fixed

**Comment by jschmid1** [comment]
done via #904
